File: typescript/src/http/client-handlers.ts
/**
 * Client-facing handlers.
 *
 * Five endpoints per `openapi/licensing-admin.yaml`:
 *
 *   GET  /health      — liveness + build info (public)
 *   POST /activate    — license-key → signed LIC1 token (public)
 *   POST /refresh     — swap a valid token for a fresher one (public)
 *   POST /heartbeat   — periodic liveness; MAY rotate token (public)
 *   POST /deactivate  — release a seat by token (public, body-authenticated)
 *
 * Every path is registered as `public: true` so the bearer-auth middleware
 * skips them. Authentication on client endpoints is body-derived (license
 * key, signed token) rather than header-derived (admin bearer).
 */

import {
  AlgorithmRegistry,
  decodeUnverified,
  findLicenseByKey,
  issueToken,
  KeyAlgBindings,
  type KeyRecord,
  LicensingError,
  registerUsage,
  revokeUsage,
  verify,
} from '../index.ts';
import type { ClientHandlerContext } from './context.ts';
import { err, errFromLicensing, noContent, ok } from './envelope.ts';
import type { Route } from './router.ts';
import type { HandlerRequest, HandlerResponse, JsonValue } from './types.ts';
import {
  optionalObject,
  requireFingerprint,
  requireJsonObject,
  requireString,
} from './validation.ts';

// Core does not export `unixSeconds`; inline the 5-line equivalent. Kept
// strict: any malformed ISO-8601 returns NaN, which callers MUST guard
// against (we only use this after a successful `clock.nowIso()` so it's safe).
function unixSeconds(iso: string): number {
  return Math.floor(new Date(iso).getTime() / 1000);
}

function isoOf(unixSec: number): string {
  return new Date(unixSec * 1000).toISOString();
}

/**
 * Verify a client-presented LIC1 token against the kid's public key loaded
 * from storage, then enforce nbf/exp with an optional past-exp grace.
 *
 * Ordering mirrors `lic1.verify`: header shape → kid→alg binding (one-shot
 * for this request) → signature. Only after the signature is valid do we
 * look at expiry — an attacker's unsigned token never influences the
 * expiry path.
 *
 * `allowExpiredWithinSec` lets /refresh accept a token that expired
 * recently (so a paused client can rotate) while /heartbeat and /deactivate
 * pass 0 and reject any expired token.
 *
 * This function is the single gatekeeper between "untrusted body bytes"
 * and "payload claims we act on" — the caller must never read license_id /
 * usage_id from `decodeUnverified`.
 */
async function verifyClientToken(
  ctx: ClientHandlerContext,
  token: string,
  allowExpiredWithinSec: number,
): Promise<ReturnType<typeof decodeUnverified>> {
  const header = decodeUnverified(token); // structural parse only — trust follows.

  const stored = await ctx.storage.getKeyByKid(header.header.kid);
  if (stored === null) {
    // `LicensingError` with UnknownKid → 401 via errFromLicensing.
    throw new LicensingError('UnknownKid', `unknown kid: ${header.header.kid}`, {
      kid: header.header.kid,
    });
  }

  // One-shot registry + bindings for just this (kid, alg). Building on the
  // fly avoids keeping a stale alg-binding cache across key rotation.
  const registry = new AlgorithmRegistry();
  const backend = ctx.backends.get(stored.alg);
  if (backend === undefined) {
    throw new LicensingError('UnsupportedAlgorithm', `backend missing for alg: ${stored.alg}`, {
      alg: stored.alg,
    });
  }
  registry.register(backend);

  const bindings = new KeyAlgBindings();
  bindings.bind(stored.kid, stored.alg);

  const record: KeyRecord = {
    kid: stored.kid,
    alg: stored.alg,
    publicPem: stored.public_pem,
    privatePem: null,
    raw: { privateRaw: null, publicRaw: new Uint8Array(0) }, // PEM path used.
  };

  const parts = await verify(token, {
    registry,
    bindings,
    keys: new Map([[record.kid, record]]),
  });

  const nowSec = unixSeconds(ctx.clock.nowIso());
  const nbf = parts.payload.nbf;
  if (typeof nbf === 'number' && nowSec < nbf) {
    throw new LicensingError('TokenMalformed', 'token not yet valid (nbf in future)');
  }
  const exp = parts.payload.exp;
  if (typeof exp === 'number' && nowSec >= exp + allowExpiredWithinSec) {
    // Strict `>=` matches the interop grace-table fix: a token at exactly
    // exp is already expired. Grace extends the window past exp, not before.
    throw new LicensingError('TokenExpired', 'token is expired');
  }

  return parts;
}

/**
 * Probe the issuer with a cheap storage read. Returns 200 + `{status, version,
 * time}` on success and 503 + `{status: "error", version, time}` when the
 * probe fails. Mirrors the Go `handleHealth`.
 *
 * The 503 keeps the success-envelope shape (`success: true, data: {...}`)
 * because the high-level client probes `/health` purely on HTTP status —
 * switching the envelope flag here would force the transport layer to
 * special-case parsing and blur the boundary between "issuer is unreachable"
 * and "issuer rejected the request".
 */
async function handleHealth(ctx: ClientHandlerContext): Promise<HandlerResponse> {
  const time = new Date().toISOString();
  try {
    await ctx.storage.listAudit({}, { limit: 1 });
  } catch {
    return ok({ status: 'error', version: ctx.version, time }, 503);
  }
  return ok({ status: 'ok', version: ctx.version, time });
}

async function handleActivate(
  ctx: ClientHandlerContext,
  req: HandlerRequest,
): Promise<HandlerResponse> {
  const body = requireJsonObject(req);
  if (!body.ok) return body.response;
  const licenseKey = requireString(body.value, 'license_key');
  if (!licenseKey.ok) return licenseKey.response;
  const fingerprint = requireFingerprint(body.value);
  if (!fingerprint.ok) return fingerprint.response;
  const clientMeta = optionalObject(body.value, 'client_meta');
  if (!clientMeta.ok) return clientMeta.response;

  try {
    const license = await findLicenseByKey(ctx.storage, licenseKey.value);
    if (license === null) {
      // "Activate with invalid key returns 404" — explicit code.
      return err(404, 'InvalidLicenseKey', 'license key is invalid or unknown');
    }
    // Register the usage (idempotent on fingerprint); possibly promotes
    // the license `pending → active`.
    const reg = await registerUsage(ctx.storage, ctx.clock, {
      license_id: license.id,
      fingerprint: fingerprint.value,
      ...(clientMeta.value !== null ? { client_meta: clientMeta.value } : {}),
    });
    const alg = ctx.defaultAlg ?? 'ed25519';
    const ttl = ctx.tokenTtlSec ?? 3600;
    const issued = await issueToken(ctx.storage, ctx.clock, ctx.backends, {
      license: reg.license,
      usage: reg.usage,
      ttlSeconds: ttl,
      alg,
      signingPassphrase: ctx.signingPassphrase,
      ...(ctx.forceOnlineAfter !== undefined ? { forceOnlineAfter: ctx.forceOnlineAfter } : {}),
    });
    // `refresh_recommended_at` is the 75%-of-lifetime mark: clients
    // proactively refresh when less than 25% remains.
    const refreshAt = issued.iat + Math.floor(ttl * 0.75);
    const body: Record<string, JsonValue> = {
      token: issued.token,
      expires_at: isoOf(issued.exp),
      refresh_recommended_at: isoOf(refreshAt),
    };
    // OpenAPI allows explicit null for force_online_after; surface it so
    // clients can surface a deterministic hard deadline when present.
    const parts = decodeUnverified(issued.token);
    const foa = parts.payload.force_online_after;
    if (typeof foa === 'number') {
      body.force_online_after = isoOf(foa);
    }
    return ok(body);
  } catch (e) {
    return errFromLicensing(e);
  }
}

/** Refresh re-issues a fresh token from a validly-signed one. We DO verify
 *  the incoming signature (B12) — prior to that an attacker holding any
 *  valid (license_id, usage_id) pair could mint tokens by submitting an
 *  unsigned payload. /refresh accepts recently-expired tokens so a paused
 *  client can still rotate; /heartbeat and /deactivate are strict. */
async function reissueFromToken(
  ctx: ClientHandlerContext,
  token: string,
  allowExpiredWithinSec: number,
): Promise<
  | { readonly token: string; readonly expiresAt: string; readonly foa: string | null }
  | HandlerResponse
> {
  let parts: ReturnType<typeof decodeUnverified>;
  try {
    parts = await verifyClientToken(ctx, token, allowExpiredWithinSec);
  } catch (e) {
    return errFromLicensing(e);
  }
  const licenseId = parts.payload.license_id;
  const usageId = parts.payload.usage_id;
  if (typeof licenseId !== 'string' || typeof usageId !== 'string') {
    return err(401, 'TokenMalformed', 'token missing required claims');
  }

  const license = await ctx.storage.getLicense(licenseId);
  if (license === null) return err(404, 'LicenseNotFound', `license not found: ${licenseId}`);
  const usage = await ctx.storage.getUsage(usageId);
  if (usage === null) return err(404, 'NotFound', `usage not found: ${usageId}`);
  if (usage.license_id !== license.id) {
    return err(401, 'TokenMalformed', 'token usage does not belong to token license');
  }
  if (usage.status !== 'active') {
    return err(403, 'LicenseRevoked', `usage ${usage.id} is no longer active`);
  }

  const alg = ctx.defaultAlg ?? 'ed25519';
  const ttl = ctx.tokenTtlSec ?? 3600;
  const issued = await issueToken(ctx.storage, ctx.clock, ctx.backends, {
    license,
    usage,
    ttlSeconds: ttl,
    alg,
    signingPassphrase: ctx.signingPassphrase,
    ...(ctx.forceOnlineAfter !== undefined ? { forceOnlineAfter: ctx.forceOnlineAfter } : {}),
  });
  const decoded = decodeUnverified(issued.token);
  const foa = decoded.payload.force_online_after;
  return {
    token: issued.token,
    expiresAt: isoOf(issued.exp),
    foa: typeof foa === 'number' ? isoOf(foa) : null,
  };
}

async function handleRefresh(
  ctx: ClientHandlerContext,
  req: HandlerRequest,
): Promise<HandlerResponse> {
  const body = requireJsonObject(req);
  if (!body.ok) return body.response;
  const token = requireString(body.value, 'token');
  if (!token.ok) return token.response;

  try {
    // Grace: tokens that expired within the TTL are still accepted on
    // /refresh — a client paused (sleep/network partition) must be able
    // to rotate. Beyond TTL seconds past exp, the token is too stale.
    const ttl = ctx.tokenTtlSec ?? 3600;
    const result = await reissueFromToken(ctx, token.value, ttl);
    if ('status' in result) return result; // already an error response
    const nowSec = unixSeconds(ctx.clock.nowIso());
    const refreshAt = nowSec + Math.floor(ttl * 0.75);
    const out: Record<string, JsonValue> = {
      token: result.token,
      expires_at: result.expiresAt,
      refresh_recommended_at: isoOf(refreshAt),
    };
    if (result.foa !== null) out.force_online_after = result.foa;
    return ok(out);
  } catch (e) {
    if (e instanceof LicensingError) return errFromLicensing(e);
    return err(500, 'InternalError', 'an unexpected error occurred');
  }
}

async function handleHeartbeat(
  ctx: ClientHandlerContext,
  req: HandlerRequest,
): Promise<HandlerResponse> {
  const body = requireJsonObject(req);
  if (!body.ok) return body.response;
  const token = requireString(body.value, 'token');
  if (!token.ok) return token.response;

  // Policy: heartbeat returns `{ok:true}` without rotating by default.
  // Strict token verification — no expiry grace. An expired token cannot
  // keep a seat warm. See B12.
  try {
    let parts: ReturnType<typeof decodeUnverified>;
    try {
      parts = await verifyClientToken(ctx, token.value, 0);
    } catch (e) {
      return errFromLicensing(e);
    }
    const licenseId = parts.payload.license_id;
    const usageId = parts.payload.usage_id;
    if (typeof licenseId !== 'string' || typeof usageId !== 'string') {
      return err(401, 'TokenMalformed', 'token missing required claims');
    }
    const license = await ctx.storage.getLicense(licenseId);
    if (license === null) return err(404, 'LicenseNotFound', `license not found: ${licenseId}`);
    if (license.status === 'revoked') return err(403, 'LicenseRevoked', 'license is revoked');
    if (license.status === 'suspended') return err(403, 'LicenseSuspended', 'license is suspended');
    const usage = await ctx.storage.getUsage(usageId);
    if (usage === null || usage.status !== 'active') {
      return err(403, 'LicenseRevoked', `usage ${usageId} is no longer active`);
    }
    return ok({
      ok: true,
      server_time: ctx.clock.nowIso(),
    });
  } catch (e) {
    return errFromLicensing(e);
  }
}

async function handleDeactivate(
  ctx: ClientHandlerContext,
  req: HandlerRequest,
): Promise<HandlerResponse> {
  const body = requireJsonObject(req);
  if (!body.ok) return body.response;
  const token = requireString(body.value, 'token');
  if (!token.ok) return token.response;
  const reason = requireString(body.value, 'reason');
  if (!reason.ok) return reason.response;
  if (!['user_requested', 'uninstall', 'reassign', 'other'].includes(reason.value)) {
    return err(400, 'BadRequest', `invalid reason: ${reason.value}`);
  }

  // /deactivate: strict verification. An attacker who learns a usage_id
  // must not be able to free someone else's seat by submitting an
  // unsigned or wrong-key token. See B12.
  try {
    let parts: ReturnType<typeof decodeUnverified>;
    try {
      parts = await verifyClientToken(ctx, token.value, 0);
    } catch (e) {
      return errFromLicensing(e);
    }
    const usageId = parts.payload.usage_id;
    if (typeof usageId !== 'string') {
      return err(401, 'TokenMalformed', 'token missing usage_id');
    }
    const usage = await ctx.storage.getUsage(usageId);
    if (usage === null) return err(404, 'NotFound', `usage not found: ${usageId}`);
    // Idempotent — already-revoked usages still return 204.
    if (usage.status !== 'revoked') {
      await revokeUsage(ctx.storage, ctx.clock, usageId, { actor: `client:${reason.value}` });
    }
    return noContent();
  } catch (e) {
    return errFromLicensing(e);
  }
}

/** Build the client-facing route set. Prefix is added by the caller when
 *  mounting the handler group (typical: `/api/licensing/v1`). */
export function clientRoutes(ctx: ClientHandlerContext, prefix = ''): readonly Route[] {
  const p = (path: string) => `${prefix}${path}`;
  return [
    { method: 'GET', pattern: p('/health'), handler: () => handleHealth(ctx), public: true },
    {
      method: 'POST',
      pattern: p('/activate'),
      handler: (req) => handleActivate(ctx, req),
      public: true,
    },
    {
      method: 'POST',
      pattern: p('/refresh'),
      handler: (req) => handleRefresh(ctx, req),
      public: true,
    },
    {
      method: 'POST',
      pattern: p('/heartbeat'),
      handler: (req) => handleHeartbeat(ctx, req),
      public: true,
    },
    {
      method: 'POST',
      pattern: p('/deactivate'),
      handler: (req) => handleDeactivate(ctx, req),
      public: true,
    },
  ];
}

File: licensing/http/client_handlers.go
package http

import (
	"encoding/json"
	"net/http"
	"time"

	lic "github.com/AnoRebel/licensing/licensing"
)

// ClientHandler is the http.Handler that multiplexes client-facing routes:
//
//	GET  /health
//	POST /activate
//	POST /refresh
//	POST /heartbeat
//	POST /deactivate
//
// Handlers are public (no bearer auth). They authenticate via body contents:
// license key for activate; signed token for refresh/heartbeat/deactivate.
type ClientHandler struct {
	ctx    *ClientContext
	prefix string
}

// NewClientHandler returns an http.Handler mounted at the given prefix
// (typically "/api/licensing/v1"). Unknown paths under the prefix → 404;
// wrong method for a known path → 405.
func NewClientHandler(ctx *ClientContext, prefix string) *ClientHandler {
	return &ClientHandler{ctx: ctx, prefix: prefix}
}

// ServeHTTP dispatches to the matching endpoint.
func (h *ClientHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path
	if h.prefix != "" {
		if len(path) < len(h.prefix) || path[:len(h.prefix)] != h.prefix {
			writeError(w, 404, "NotFound", "no handler for "+r.Method+" "+path)
			return
		}
		path = path[len(h.prefix):]
	}

	switch path {
	case "/health":
		if r.Method != http.MethodGet {
			writeError(w, 405, "MethodNotAllowed", "method "+r.Method+" not allowed for "+r.URL.Path)
			return
		}
		h.handleHealth(w, r)
	case "/activate":
		if r.Method != http.MethodPost {
			writeError(w, 405, "MethodNotAllowed", "method "+r.Method+" not allowed for "+r.URL.Path)
			return
		}
		h.handleActivate(w, r)
	case "/refresh":
		if r.Method != http.MethodPost {
			writeError(w, 405, "MethodNotAllowed", "method "+r.Method+" not allowed for "+r.URL.Path)
			return
		}
		h.handleRefresh(w, r)
	case "/heartbeat":
		if r.Method != http.MethodPost {
			writeError(w, 405, "MethodNotAllowed", "method "+r.Method+" not allowed for "+r.URL.Path)
			return
		}
		h.handleHeartbeat(w, r)
	case "/deactivate":
		if r.Method != http.MethodPost {
			writeError(w, 405, "MethodNotAllowed", "method "+r.Method+" not allowed for "+r.URL.Path)
			return
		}
		h.handleDeactivate(w, r)
	default:
		writeError(w, 404, "NotFound", "no handler for "+r.Method+" "+r.URL.Path)
	}
}

// ---------------- helpers ----------------

func isoOf(unixSec int64) string {
	return time.Unix(unixSec, 0).UTC().Format(time.RFC3339)
}

// verifyClientToken verifies the signature on a client-presented LIC1 token
// against the kid's public key loaded from storage, then enforces nbf/exp
// with an optional past-exp grace.
//
// Ordering mirrors lic.Verify: header shape → kid→alg binding (one-shot, per
// this request) → signature. Only after the signature is valid do we look at
// timing — an attacker's unsigned token never influences the expiry path.
//
// allowExpiredWithin lets /refresh accept a token that expired recently (so a
// legitimately-paused client can still rotate its token) while /heartbeat and
// /deactivate pass 0 and refuse any expired token outright.
//
// This function is the single gatekeeper between "untrusted bytes in the
// request body" and "payload claims we act on" — the caller must never read
// license_id/usage_id from DecodeUnverified.
func verifyClientToken(ctx *ClientContext, token string, allowExpiredWithin time.Duration) (lic.LIC1DecodedParts, error) {
	var zero lic.LIC1DecodedParts

	// Peek at the header to learn which kid signed this. DecodeUnverified
	// only parses structure; trust is established by the Verify call below.
	head, err := lic.DecodeUnverified(token)
	if err != nil {
		return zero, err
	}

	record, err := ctx.Storage.GetKeyByKid(head.Header.Kid)
	if err != nil {
		return zero, err
	}
	if record == nil {
		return zero, lic.NewError(lic.CodeUnknownKid,
			"unknown kid: "+head.Header.Kid,
			map[string]any{"kid": head.Header.Kid})
	}

	// One-shot bindings table: only this (kid, alg) pair is accepted. If the
	// header's alg disagrees with the stored record, Expect returns
	// AlgorithmMismatch before any backend touches the signature — the
	// algorithm-confusion defense we document in docs/security.md.
	bindings := lic.NewKeyAlgBindings()
	if err := bindings.Bind(record.Kid, record.Alg); err != nil {
		return zero, err
	}

	parts, err := lic.Verify(token, lic.VerifyOptions{
		Registry: ctx.Backends,
		Bindings: bindings,
		Keys: map[string]lic.KeyRecord{
			record.Kid: {
				Kid: record.Kid,
				Alg: record.Alg,
				Pem: lic.PemKeyMaterial{PublicPem: record.PublicPem},
			},
		},
	})
	if err != nil {
		return zero, err
	}

	// nbf/exp are written as unix seconds by token_service.go; after JSON
	// decode they come back as float64.
	now, err := time.Parse(time.RFC3339, ctx.Clock.NowISO())
	if err != nil {
		now = time.Now().UTC()
	}
	nowSec := now.Unix()

	if nbf, ok := parts.Payload["nbf"].(float64); ok {
		if nowSec < int64(nbf) {
			return zero, lic.NewError(lic.CodeTokenMalformed,
				"token not yet valid (nbf in future)", nil)
		}
	}
	if exp, ok := parts.Payload["exp"].(float64); ok {
		// Strict `<` matches the interop grace-table fix: a token at exactly
		// exp is already expired. Grace extends the acceptance window past
		// exp by allowExpiredWithin, not before it.
		if nowSec >= int64(exp)+int64(allowExpiredWithin.Seconds()) {
			return zero, lic.NewError(lic.CodeTokenExpired,
				"token is expired", nil)
		}
	}

	return parts, nil
}

// maxRequestBodyBytes caps request-body size to prevent memory-exhaustion
// DoS via giant JSON payloads. LIC1 tokens are well under 4 KiB; the
// largest admin request (key import with encrypted PKCS#8 PEM) still sits
// comfortably under 1 MiB. 1 MiB leaves generous headroom for UTF-8
// expansion and future fields without opening a hole.
const maxRequestBodyBytes int64 = 1 << 20 // 1 MiB

// decodeBody reads the JSON body into out. On failure writes 400 BadRequest
// and returns false; callers should return immediately. The body is capped
// at maxRequestBodyBytes — oversize payloads surface as a decode error and
// get a 400 response.
func decodeBody(w http.ResponseWriter, r *http.Request, out any) bool {
	if r.Body == nil {
		writeError(w, 400, "BadRequest", "request body is required")
		return false
	}
	r.Body = http.MaxBytesReader(w, r.Body, maxRequestBodyBytes)
	defer r.Body.Close()
	dec := json.NewDecoder(r.Body)
	dec.DisallowUnknownFields()
	if err := dec.Decode(out); err != nil {
		writeError(w, 400, "BadRequest", "invalid JSON body: "+err.Error())
		return false
	}
	return true
}

// ---------------- /health ----------------

// handleHealth probes storage with a cheap read to verify the issuer is
// actually capable of serving requests (not just listening on the port).
// Returns 200 with `{status, version, time}` on success and 503 with
// `{status: "error", version, time}` when the storage probe fails. The
// probe is intentionally bounded — a `ListAudit` with limit 1 — so it
// catches connection drops and permission issues without paying for a
// full table scan or a write.
//
// The 503 response is still wrapped in the standard success envelope
// (`success: true, data: {status: "error", ...}`) because the high-level
// client uses `/health` as a *liveness* signal, not as a request that can
// itself fail with a typed protocol error. Switching the envelope's
// `success` flag here would force the client to special-case parsing and
// blur the boundary between "issuer is unreachable" and "issuer rejected
// the request".
func (h *ClientHandler) handleHealth(w http.ResponseWriter, _ *http.Request) {
	now := time.Now().UTC().Format(time.RFC3339)
	if h.ctx.Storage == nil {
		writeOKStatus(w, http.StatusServiceUnavailable, map[string]any{
			"status":  "error",
			"version": h.ctx.Version,
			"time":    now,
		})
		return
	}
	if _, err := h.ctx.Storage.ListAudit(lic.AuditLogFilter{}, lic.PageRequest{Limit: 1}); err != nil {
		writeOKStatus(w, http.StatusServiceUnavailable, map[string]any{
			"status":  "error",
			"version": h.ctx.Version,
			"time":    now,
		})
		return
	}
	writeOK(w, map[string]any{
		"status":  "ok",
		"version": h.ctx.Version,
		"time":    now,
	})
}

// ---------------- /activate ----------------

type activateBody struct {
	ClientMeta  map[string]any `json:"client_meta,omitempty"`
	LicenseKey  string         `json:"license_key"`
	Fingerprint string         `json:"fingerprint"`
}

func (h *ClientHandler) handleActivate(w http.ResponseWriter, r *http.Request) {
	var body activateBody
	if !decodeBody(w, r, &body) {
		return
	}
	if body.LicenseKey == "" {
		writeError(w, 400, "BadRequest", "license_key is required")
		return
	}
	if body.Fingerprint == "" {
		writeError(w, 400, "BadRequest", "fingerprint is required")
		return
	}

	license, err := lic.FindLicenseByKey(h.ctx.Storage, body.LicenseKey)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}
	if license == nil {
		writeError(w, 404, "InvalidLicenseKey", "license key is invalid or unknown")
		return
	}

	reg, err := lic.RegisterUsage(h.ctx.Storage, h.ctx.Clock, lic.RegisterUsageInput{
		LicenseID:   license.ID,
		Fingerprint: body.Fingerprint,
		ClientMeta:  body.ClientMeta,
	}, lic.RegisterUsageOptions{Actor: "http:activate"})
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}

	issued, err := h.issueFor(reg.License, reg.Usage)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}

	resp := map[string]any{
		"token":                  issued.Token,
		"expires_at":             isoOf(issued.Exp),
		"refresh_recommended_at": isoOf(issued.Iat + int64(float64(h.ctx.ttl())*0.75)),
	}
	if foa := extractForceOnlineAfter(issued.Token); foa != nil {
		resp["force_online_after"] = isoOf(*foa)
	}
	writeOK(w, resp)
}

// ---------------- /refresh ----------------

type refreshBody struct {
	Token string `json:"token"`
}

func (h *ClientHandler) handleRefresh(w http.ResponseWriter, r *http.Request) {
	var body refreshBody
	if !decodeBody(w, r, &body) {
		return
	}
	if body.Token == "" {
		writeError(w, 400, "BadRequest", "token is required")
		return
	}

	// /refresh accepts tokens whose exp just passed — a legitimate client
	// paused (sleep/suspend, network partition) should still be able to
	// rotate. The grace is capped at the configured TTL so a long-expired
	// token (days/weeks) can never refresh. Signature + kid binding are
	// still strictly enforced.
	refreshGrace := time.Duration(h.ctx.ttl()) * time.Second
	parts, err := verifyClientToken(h.ctx, body.Token, refreshGrace)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}
	licenseID, _ := parts.Payload["license_id"].(string)
	usageID, _ := parts.Payload["usage_id"].(string)
	if licenseID == "" || usageID == "" {
		writeError(w, 401, "TokenMalformed", "token missing required claims")
		return
	}

	var (
		license *lic.License
		usage   *lic.LicenseUsage
	)
	txErr := h.ctx.Storage.WithTransaction(func(tx lic.StorageTx) error {
		l, err := tx.GetLicense(licenseID)
		if err != nil {
			return err
		}
		if l == nil {
			return lic.NewError(lic.CodeLicenseNotFound, "license not found: "+licenseID, nil)
		}
		u, err := tx.GetUsage(usageID)
		if err != nil {
			return err
		}
		if u == nil {
			return lic.NewError(lic.CodeLicenseNotFound, "usage not found: "+usageID, nil)
		}
		if u.LicenseID != l.ID {
			return lic.NewError(lic.CodeTokenMalformed, "token usage does not belong to token license", nil)
		}
		if u.Status != lic.UsageStatusActive {
			return lic.NewError(lic.CodeLicenseRevoked, "usage "+u.ID+" is no longer active", nil)
		}
		license = l
		usage = u
		return nil
	})
	if txErr != nil {
		writeErrorFromLicensing(w, txErr)
		return
	}

	issued, err := h.issueFor(license, usage)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}

	resp := map[string]any{
		"token":                  issued.Token,
		"expires_at":             isoOf(issued.Exp),
		"refresh_recommended_at": isoOf(issued.Iat + int64(float64(h.ctx.ttl())*0.75)),
	}
	if foa := extractForceOnlineAfter(issued.Token); foa != nil {
		resp["force_online_after"] = isoOf(*foa)
	}
	writeOK(w, resp)
}

// ---------------- /heartbeat ----------------

type heartbeatBody struct {
	Token string `json:"token"`
}

func (h *ClientHandler) handleHeartbeat(w http.ResponseWriter, r *http.Request) {
	var body heartbeatBody
	if !decodeBody(w, r, &body) {
		return
	}
	if body.Token == "" {
		writeError(w, 400, "BadRequest", "token is required")
		return
	}

	// /heartbeat: strict verification, no expiry grace. An expired token
	// cannot keep a seat warm.
	parts, err := verifyClientToken(h.ctx, body.Token, 0)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}
	licenseID, _ := parts.Payload["license_id"].(string)
	usageID, _ := parts.Payload["usage_id"].(string)
	if licenseID == "" || usageID == "" {
		writeError(w, 401, "TokenMalformed", "token missing required claims")
		return
	}

	txErr := h.ctx.Storage.WithTransaction(func(tx lic.StorageTx) error {
		l, err := tx.GetLicense(licenseID)
		if err != nil {
			return err
		}
		if l == nil {
			return lic.NewError(lic.CodeLicenseNotFound, "license not found: "+licenseID, nil)
		}
		if l.Status == lic.LicenseStatusRevoked {
			return lic.NewError(lic.CodeLicenseRevoked, "license is revoked", nil)
		}
		if l.Status == lic.LicenseStatusSuspended {
			return lic.NewError(lic.CodeLicenseSuspended, "license is suspended", nil)
		}
		u, err := tx.GetUsage(usageID)
		if err != nil {
			return err
		}
		if u == nil || u.Status != lic.UsageStatusActive {
			return lic.NewError(lic.CodeLicenseRevoked, "usage "+usageID+" is no longer active", nil)
		}
		return nil
	})
	if txErr != nil {
		writeErrorFromLicensing(w, txErr)
		return
	}

	writeOK(w, map[string]any{
		"ok":          true,
		"server_time": h.ctx.Clock.NowISO(),
	})
}

// ---------------- /deactivate ----------------

type deactivateBody struct {
	Token  string `json:"token"`
	Reason string `json:"reason"`
}

var allowedDeactivateReasons = map[string]bool{
	"user_requested": true,
	"uninstall":      true,
	"reassign":       true,
	"other":          true,
}

func (h *ClientHandler) handleDeactivate(w http.ResponseWriter, r *http.Request) {
	var body deactivateBody
	if !decodeBody(w, r, &body) {
		return
	}
	if body.Token == "" {
		writeError(w, 400, "BadRequest", "token is required")
		return
	}
	if !allowedDeactivateReasons[body.Reason] {
		writeError(w, 400, "BadRequest", "invalid reason: "+body.Reason)
		return
	}

	// /deactivate: strict verification. An attacker who learns a usage_id
	// must not be able to free someone else's seat by submitting an
	// unsigned or wrong-key token.
	parts, err := verifyClientToken(h.ctx, body.Token, 0)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}
	usageID, _ := parts.Payload["usage_id"].(string)
	if usageID == "" {
		writeError(w, 401, "TokenMalformed", "token missing usage_id")
		return
	}

	usage, err := h.ctx.Storage.GetUsage(usageID)
	if err != nil {
		writeErrorFromLicensing(w, err)
		return
	}
	if usage == nil {
		writeError(w, 404, "NotFound", "usage not found: "+usageID)
		return
	}
	if usage.Status != lic.UsageStatusRevoked {
		if _, err := lic.RevokeUsage(h.ctx.Storage, h.ctx.Clock, usageID, lic.RevokeUsageOptions{
			Actor: "client:" + body.Reason,
		}); err != nil {
			writeErrorFromLicensing(w, err)
			return
		}
	}
	writeNoContent(w)
}

// ---------------- helpers ----------------

func (h *ClientHandler) issueFor(license *lic.License, usage *lic.LicenseUsage) (*lic.IssueTokenResult, error) {
	input := lic.IssueTokenInput{
		License:           license,
		Usage:             usage,
		TTLSeconds:        h.ctx.ttl(),
		Alg:               h.ctx.alg(),
		SigningPassphrase: h.ctx.SigningPassphrase,
	}
	if h.ctx.ForceOnlineAfter != nil {
		v := *h.ctx.ForceOnlineAfter
		input.ForceOnlineAfter = lic.OptIntOverride{Set: true, Value: &v}
	}
	return lic.IssueToken(h.ctx.Storage, h.ctx.Clock, h.ctx.Backends, input)
}

// extractForceOnlineAfter peeks at a freshly-issued token and returns its
// force_online_after claim if present. Nil if absent or malformed.
func extractForceOnlineAfter(token string) *int64 {
	parts, err := lic.DecodeUnverified(token)
	if err != nil {
		return nil
	}
	v, ok := parts.Payload["force_online_after"]
	if !ok {
		return nil
	}
	switch x := v.(type) {
	case float64:
		i := int64(x)
		return &i
	case json.Number:
		if i, err := x.Int64(); err == nil {
			return &i
		}
	case int64:
		return &x
	}
	return nil
}
